src/server/game/Entities/Creature/Creature.cpp

```diff
@@ -1206,11 +1197,13 @@ bool Creature::isCanInteractWithBattleMaster(Player* player, bool msg) const
     return true;
 }
 
-bool Creature::isCanTrainingAndResetTalentsOf(Player* player) const
+bool Creature::CanResetTalents(Player* player) const
 {
-    return player->getLevel() >= 10
-        && GetCreatureTemplate()->trainer_type == TRAINER_TYPE_CLASS
-        && player->getClass() == GetCreatureTemplate()->trainer_class;
+    Trainer::Trainer const* trainer = sObjectMgr->GetTrainer(GetEntry());
+    if (!trainer)
+        return false;
+
+    return player->getLevel() >= 10 && trainer->IsTrainerValidForPlayer(player);
```

Comment:
I think this will return true for every professions trainers as well
`&& trainer->GetTrainerType() == Trainer::Type::Class;`